src/code.rs

```diff
@@ -153,32 +168,39 @@ impl<'a> Struct<'a> {
 
     /// Assemble all derives for the struct
     fn attr_derive(&self) -> String {
-        format!("#[derive(Debug, {derive_serde}Clone, Queryable, Insertable{derive_aschangeset}{derive_identifiable}{derive_associations}{derive_selectable}{derive_default})]",
-                derive_selectable = match self.ty {
-                    StructType::Read => { ", Selectable" }
-                    _ => { "" }
-                },
-                derive_associations = match self.ty {
-                    StructType::Read => {
-                        if !self.table.foreign_keys.is_empty() { ", Associations" } else { "" }
-                    }
-                    _ => { "" }
-                },
-                derive_identifiable = match self.ty {
-                    StructType::Read => {
-                        if !self.table.foreign_keys.is_empty() { ", Identifiable" } else { "" }
-                    }
-                    _ => { "" }
-                },
-                derive_aschangeset = if self.fields().iter().all(|f| self.table.primary_key_column_names().contains(&f.name)) {""} else { ", AsChangeset" },
-                derive_default = match self.ty {
-                    StructType::Update => { ", Default" }
-                    _ => { "" }
-                },
-                derive_serde = if self.config.table(&self.table.name.to_string()).get_serde() {
-                    "Serialize, Deserialize, "
-                } else { "" }
-        )
+        let mut derives_vec = Vec::with_capacity(10);
+        // Default derives that exist on every struct
+        derives_vec.extend_from_slice(&[derives::DEBUG, derives::CLONE]);
+
+        if self.config.table(&self.table.name.to_string()).get_serde() {
+            derives_vec.extend_from_slice(&[derives::SERIALIZE, derives::DESERIALIZE]);
+        }
+
+        match self.ty {
+            StructType::Read => {
+                // derives at always exist, regardless of extra conditions
+                derives_vec.extend_from_slice(&[derives::QUERYABLE, derives::SELECTABLE]);
+
+                if !self.table.foreign_keys.is_empty() {
+                    derives_vec.extend_from_slice(&[derives::ASSOCIATIONS, derives::IDENTIFIABLE]);
+                }
+            }
+            StructType::Update => {
+                // TODO: is this still really necessary?
+                if !self
+                    .fields()
+                    .iter()
+                    .all(|f| self.table.primary_key_column_names().contains(&f.name))
+                {
+                    derives_vec.push(derives::ASCHANGESET);
+                }
```

Comment:
So, this is for a special case where all fields of a table are primary keys. If that's the case, we assume it's an insert/delete-only table.

Adding a `AsChangeset` derive to a table with only primary_key columns will result in a compilation error (because there are no columns to update).

This is from memory, but I believe it's still the case with the latest diesel version and the way our code is generated. I just can't remember if it was a diesel compilation error (on the struct) or the way we've written the update() method (in our generated function).